src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+
+      val fastqRecords                        = input.next()
+      val sampleBarcodeReadBases: Seq[String] = getSampleBarcode(fastqRecords).filter(_.nonEmpty)
+      val sampleOrdinal: Int                  = getSampleOrdinalFromSampleBarcode(sampleBarcodeReadBases.mkString, sampleSheet).getOrElse(headers.length)
+      val header: SAMFileHeader               = headers(sampleOrdinal-1)
+      val sampleId: String                    = header.getReadGroups.get(0).getId
+      val molecularBarcodes: Seq[String]      = getMolecularBarcodes(fastqRecords).filter(_.nonEmpty)
+      val pairedEnd: Boolean                  = readStructures.count(_.template.nonEmpty) > 1
+
+      val records = fastqRecords.zipWithIndex.flatMap { case (record, recordIndex) =>
+        val readStructure = templateReadStructures(recordIndex)
+        if (readStructure.isEmpty) {
+          None
+        }
+        else {
+          if (1 < recordIndex) throw new IllegalArgumentException("Found more than two ends with template bases")
+          val samRecord: SAMRecord = makeSAMRecord(header, record, readStructure, sampleId, pairedEnd, recordIndex == 0)
+          
+          // set attributes: sample and molecular barcodes
+          samRecord.setAttribute("BC", sampleBarcodeReadBases.mkString(SampleBarcodeDelimiter))
+          val umiValue = molecularBarcodes.mkString(SampleBarcodeDelimiter)
+          if (umiValue.nonEmpty) samRecord.setAttribute(umiTag, molecularBarcodes.mkString(SampleBarcodeDelimiter))
+
+          if (multiUmiTags.nonEmpty) {
+            val numMolecularBarcodeSegments = getNumMolecularBarcodes(fastqRecords)
+            if (numMolecularBarcodeSegments == multiUmiTags.length) {
+              multiUmiTags.zip(molecularBarcodes).foreach { case (tag, barcode) => samRecord.setAttribute(tag, barcode) }
+            }
+            else if (multiUmiTags.length == 1) {
+              samRecord.setAttribute(multiUmiTags.head, molecularBarcodes.mkString(SampleBarcodeDelimiter))
+            }
+            else { // more than one tag, but not the same as the number of molecular barcodes
+              unreachable(s"Did not find only one SAM tag or the same # of SAM tags for the molecular barcodes (${multiUmiTags.length} != $numMolecularBarcodeSegments)")
+            }
+          }
+          Some(samRecord)
+        }
+      }
+      AssignedRecords(sampleOrdinal=sampleOrdinal, records=records)
+    }
+
+    def getBarcodeMetrics: Iterable[SampleBarcodeMetric] = {
+      SampleBarcodeMetric.finalizeMetrics(this.barcodeToMetrics, this.noMatchBarcode)
+      this.barcodeToMetrics.values
+    }
+
+    private def noMatchBarcodeMetric: SampleBarcodeMetric = this.barcodeToMetrics(noMatchBarcode)
+
+    /** Gets the # of mismatches for the given sample and observed sample barcode read bases. */
+    private def sampleOrdinalAndMismatches(sample: Sample, sampleBarcodeReadBases: String): (Int, Int) = {
```

Comment:
Return type needs to be documented.  Would suggest also either making it a case class, or adding a type def for `Ordinal = Int` and using that everywhere to make it clearer.